plugins/crypto/mbedtls/certificategroup.c

```diff
@@ -132,6 +134,119 @@ MemoryCertStore_getRejectedList(UA_CertificateGroup *certGroup, UA_ByteString **
     return retval;
 }
 
+static UA_StatusCode
+checkCrlMatch(const UA_ByteString *certificate, const UA_ByteString *crl) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    mbedtls_x509_crt cert;
+    mbedtls_x509_crt_init(&cert);
+
+    mbedtls_x509_crl certRevocationList;
+    mbedtls_x509_crl_init(&certRevocationList);
+
+    UA_ByteString data1 = UA_mbedTLS_CopyDataFormatAware(certificate);
+    UA_ByteString data2 = UA_mbedTLS_CopyDataFormatAware(crl);
+
+    int mbedErr = mbedtls_x509_crt_parse(&cert, data1.data, data1.length);
+    if(mbedErr) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    mbedErr = mbedtls_x509_crl_parse(&certRevocationList, data2.data, data2.length);
+    if(mbedErr) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    /* Check if the certificate is a CA certificate.
+     * Only a CA certificate can have a CRL. */
+    if(!mbedtlsCheckCA(&cert)) {
+        retval = UA_STATUSCODE_BADCERTIFICATEUSENOTALLOWED;
+        goto cleanup;
+    }
+
+    char certSubject[MBEDTLS_X509_MAX_DN_NAME_SIZE];
+    char crlIssuer[MBEDTLS_X509_MAX_DN_NAME_SIZE];
+
+    mbedtls_x509_dn_gets(certSubject, sizeof(certSubject), &cert.subject);
+    mbedtls_x509_dn_gets(crlIssuer, sizeof(crlIssuer), &certRevocationList.issuer);
+
+    if(strncmp(certSubject, crlIssuer, MBEDTLS_X509_MAX_DN_NAME_SIZE) == 0) {
+        retval = UA_STATUSCODE_GOOD;
+    } else {
+        retval = UA_STATUSCODE_BADNOMATCH;
+    }
+
+cleanup:
+    mbedtls_x509_crt_free(&cert);
+    mbedtls_x509_crl_free(&certRevocationList);
+    UA_ByteString_clear(&data1);
+    UA_ByteString_clear(&data2);
+
+    return retval;
+}
+
+static UA_StatusCode
+checkCrlList(UA_CertificateGroup *certGroup, const UA_ByteString *certificate,
+             const UA_ByteString *crlList, const size_t crlListSize,
+             UA_ByteString **crls, size_t *crlsSize) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_Boolean foundMatch = false;
+    for(size_t i = 0; i < crlListSize; i++) {
+        retval = checkCrlMatch(certificate, &crlList[i]);
```

Comment:
You parse and checkCA the certificate in every iteration.
You can do it only once at the beginning.